File: sdk/apple/Sources/Qenlo/Qenlo.swift
import CQenlo
import Foundation

public struct QenloError: Error, Equatable, Sendable, CustomStringConvertible {
    public let message: String
    public var description: String { message }
    public init(_ message: String) { self.message = message }
}

public struct QenloRecord: Sendable, Equatable {
    public let id: UInt64
    public let userID: UInt64
    public let timestamp: Int64
    public let vector: [Float]
    public init(id: UInt64, userID: UInt64, timestamp: Int64, vector: [Float]) {
        self.id = id; self.userID = userID; self.timestamp = timestamp; self.vector = vector
    }
}

public struct QenloFilter: Sendable, Equatable {
    public let userID: UInt64?
    public let timestampLower: Int64?
    public let timestampUpper: Int64?
    public init(userID: UInt64? = nil, timestampLower: Int64? = nil, timestampUpper: Int64? = nil) {
        self.userID = userID; self.timestampLower = timestampLower; self.timestampUpper = timestampUpper
    }
}

public struct QenloSearchResult: Codable, Sendable, Equatable {
    public let id: UInt64
    public let distance: Float
}

public struct QenloExecutionReport: Sendable, Equatable {
    public let operationID: UInt64
    public let requestedBackend: String
    public let actualBackend: String
    public let algorithm: String
    public let filterExecution: String
    public let indexGeneration: UInt64
    public let rebuilt: Bool
    public let routingReason: String?
    public let fallbackReason: String?
    public let totalDurationNanoseconds: UInt64
    public let lockWaitNanoseconds: UInt64
    public let eligibleRows: UInt64?
    public let uploadBytes: UInt64?
    public let readbackBytes: UInt64?
    public let allocationBytes: UInt64?
    public let dispatchCount: UInt32?
    public let candidates: UInt64?
    public let batchSize: Int
}

public struct QenloSearchResponse: Sendable, Equatable {
    public let results: [QenloSearchResult]
    public let report: QenloExecutionReport
}

public struct QenloCollectionStats: Sendable, Equatable {
    public let dimension: Int
    public let rows: Int
    public let liveRows: Int
    public let generation: UInt64
    public let preparedGeneration: UInt64?
    public let durableGeneration: UInt64?
    public let recoveredInterruptedWrite: Bool
    public let closed: Bool
}

private struct WireHit: Decodable { let id: String; let distance: Float }
private struct WireReport: Decodable {
    let operationID, indexGeneration, totalDurationNanoseconds, lockWaitNanoseconds: String
    let requestedBackend, actualBackend, algorithm, filterExecution: String
    let rebuilt: Bool
    let routingReason, fallbackReason, eligibleRows, uploadBytes, readbackBytes, allocationBytes, candidates: String?
    let dispatchCount: UInt32?
    let batchSize: Int
    enum CodingKeys: String, CodingKey {
        case operationID = "operation_id", requestedBackend = "requested_backend", actualBackend = "actual_backend", algorithm
        case filterExecution = "filter_execution", indexGeneration = "index_generation", rebuilt
        case routingReason = "routing_reason", fallbackReason = "fallback_reason"
        case totalDurationNanoseconds = "total_duration_ns", lockWaitNanoseconds = "lock_wait_ns"
        case eligibleRows = "eligible_rows", uploadBytes = "upload_bytes", readbackBytes = "readback_bytes"
        case allocationBytes = "allocation_bytes", dispatchCount = "dispatch_count", candidates, batchSize = "batch_size"
    }
}
private struct WireSearch: Decodable { let results: [WireHit]; let report: WireReport }
private struct WireStats: Decodable {
    let dimension, rows, liveRows: Int
    let generation: String
    let preparedGeneration, durableGeneration: String?
    let recoveredInterruptedWrite, closed: Bool
    enum CodingKeys: String, CodingKey {
        case dimension, rows, liveRows = "live_rows", generation, preparedGeneration = "prepared_generation"
        case durableGeneration = "durable_generation", recoveredInterruptedWrite = "recovered_interrupted_write", closed
    }
}

public final class QenloCollection: @unchecked Sendable {
    public let dimension: Int
    private let lock = NSRecursiveLock()
    private var handle: OpaquePointer?

    public init(memoryDimension dimension: Int) throws {
        guard dimension > 0 else { throw QenloError("dimension must be positive") }
        self.dimension = dimension
        self.handle = qenlo_collection_new(dimension)
        guard handle != nil else { throw QenloError(Self.lastError()) }
    }

    public init(create path: URL, dimension: Int) throws {
        guard dimension > 0 else { throw QenloError("dimension must be positive") }
        self.dimension = dimension
        self.handle = path.path.withCString { qenlo_collection_create($0, dimension) }
        guard handle != nil else { throw QenloError(Self.lastError()) }
    }

    public init(open path: URL, dimension: Int) throws {
        guard dimension > 0 else { throw QenloError("dimension must be positive") }
        self.dimension = dimension
        self.handle = path.path.withCString { qenlo_collection_open($0, dimension) }
        guard handle != nil else { throw QenloError(Self.lastError()) }
    }

    deinit { try? close() }

    private static func take(_ pointer: UnsafeMutablePointer<CChar>?) throws -> String {
        guard let pointer else { throw QenloError(lastError()) }
        defer { qenlo_string_free(pointer) }
        return String(cString: pointer)
    }

    private static func lastError() -> String {
        guard let pointer = qenlo_last_error() else { return "unknown Qenlo native error" }
        defer { qenlo_string_free(pointer) }
        return String(cString: pointer)
    }

    private func withHandle<T>(_ body: (OpaquePointer) throws -> T) throws -> T {
        lock.lock(); defer { lock.unlock() }
        guard let handle else { throw QenloError("collection is closed") }
        return try body(handle)
    }

    private func checked(_ status: Int32) throws { if status != 0 { throw QenloError(Self.lastError()) } }
    private func checkedVector(_ vector: [Float]) throws -> [Float] {
        guard vector.count == dimension else { throw QenloError("expected vector dimension \(dimension), got \(vector.count)") }
        return vector
    }

    public func add(_ record: QenloRecord) throws {
        let vector = try checkedVector(record.vector)
        try withHandle { handle in try vector.withUnsafeBufferPointer { try checked(qenlo_add(handle, record.id, record.userID, record.timestamp, $0.baseAddress, vector.count)) } }
    }

    public func addBatch(_ records: [QenloRecord]) throws {
        guard !records.isEmpty else { return }
        let ids = records.map(\.id), users = records.map(\.userID), timestamps = records.map(\.timestamp)
        let vectors = try records.flatMap { try checkedVector($0.vector) }
        try withHandle { handle in
            try ids.withUnsafeBufferPointer { ids in try users.withUnsafeBufferPointer { users in try timestamps.withUnsafeBufferPointer { timestamps in try vectors.withUnsafeBufferPointer { vectors in
                try checked(qenlo_add_batch(handle, ids.baseAddress, users.baseAddress, timestamps.baseAddress, vectors.baseAddress, records.count, dimension))
            } } } }
        }
    }

    public func delete(_ id: UInt64) throws { try withHandle { try checked(qenlo_delete($0, id)) } }
    public func deleteBatch(_ ids: [UInt64]) throws {
        guard !ids.isEmpty else { return }
        try withHandle { handle in try ids.withUnsafeBufferPointer { try checked(qenlo_delete_batch(handle, $0.baseAddress, ids.count)) } }
    }

    public func search(_ query: [Float], filter: QenloFilter = .init(), k: Int = 10) throws -> QenloSearchResponse {
        guard (1...64).contains(k) else { throw QenloError("k must be in 1...64") }
        let query = try checkedVector(query)
        return try withHandle { handle in
            let json = try query.withUnsafeBufferPointer { values in
                try Self.take(qenlo_search(handle, values.baseAddress, query.count, filter.userID != nil, filter.userID ?? 0, filter.timestampLower != nil, filter.timestampLower ?? 0, filter.timestampUpper != nil, filter.timestampUpper ?? 0, k))
            }
            let wire = try JSONDecoder().decode(WireSearch.self, from: Data(json.utf8))
            guard let operation = UInt64(wire.report.operationID), let generation = UInt64(wire.report.indexGeneration), let total = UInt64(wire.report.totalDurationNanoseconds), let wait = UInt64(wire.report.lockWaitNanoseconds) else { throw QenloError("invalid 64-bit telemetry from native library") }
            return QenloSearchResponse(
                results: try wire.results.map { guard let id = UInt64($0.id) else { throw QenloError("invalid result ID") }; return .init(id: id, distance: $0.distance) },
                report: .init(operationID: operation, requestedBackend: wire.report.requestedBackend, actualBackend: wire.report.actualBackend, algorithm: wire.report.algorithm, filterExecution: wire.report.filterExecution, indexGeneration: generation, rebuilt: wire.report.rebuilt, routingReason: wire.report.routingReason, fallbackReason: wire.report.fallbackReason, totalDurationNanoseconds: total, lockWaitNanoseconds: wait, eligibleRows: wire.report.eligibleRows.flatMap(UInt64.init), uploadBytes: wire.report.uploadBytes.flatMap(UInt64.init), readbackBytes: wire.report.readbackBytes.flatMap(UInt64.init), allocationBytes: wire.report.allocationBytes.flatMap(UInt64.init), dispatchCount: wire.report.dispatchCount, candidates: wire.report.candidates.flatMap(UInt64.init), batchSize: wire.report.batchSize)
            )
        }
    }

    public func stats() throws -> QenloCollectionStats {
        try withHandle { handle in
            let wire = try JSONDecoder().decode(WireStats.self, from: Data(try Self.take(qenlo_stats(handle)).utf8))
            guard let generation = UInt64(wire.generation) else { throw QenloError("invalid generation") }
            return .init(dimension: wire.dimension, rows: wire.rows, liveRows: wire.liveRows, generation: generation, preparedGeneration: wire.preparedGeneration.flatMap(UInt64.init), durableGeneration: wire.durableGeneration.flatMap(UInt64.init), recoveredInterruptedWrite: wire.recoveredInterruptedWrite, closed: wire.closed)
        }
    }

    public func flush() throws { try withHandle { try checked(qenlo_flush($0)) } }

    public func close() throws {
        lock.lock(); defer { lock.unlock() }
        guard let value = handle else { return }
        handle = nil
        let status = qenlo_close(value)
        qenlo_collection_free(value)
        try checked(status)
    }
}

File: sdk/kotlin/src/main/kotlin/dev/qenlo/Qenlo.kt
package dev.qenlo

import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.NativeLibrary
import com.sun.jna.Pointer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.nio.file.Path
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write

/** A validation, lifecycle, storage, or native execution failure. */
public class QenloException(message: String) : RuntimeException(message)

/** Requested execution policy for collection searches. */
public enum class ExecutionMode(internal val nativeValue: Int) {
    CPU(0),
    AUTOMATIC(1),
    GPU_REQUIRED(2),
}

/** Eligibility preparation strategy used by the portable GPU backend. */
public enum class GpuFilterMode(internal val nativeValue: Int) {
    CPU_MASK(0),
    CPU_ELIGIBLE_ROWS(1),
    GPU_PREDICATE(2),
}

/** Native execution and allocation settings for a collection. */
public data class CollectionOptions(
    public val backend: ExecutionMode = ExecutionMode.CPU,
    public val gpuFilterMode: GpuFilterMode = GpuFilterMode.GPU_PREDICATE,
    public val gpuAllocationBudgetBytes: ULong = 512uL * 1024uL * 1024uL,
) {
    init { require(gpuAllocationBudgetBytes > 0uL) { "gpu allocation budget must be positive" } }
}

/** One canonical vector and its filterable metadata. */
public data class Record(
    public val id: ULong,
    public val userId: ULong,
    public val timestamp: Long,
    public val vector: FloatArray,
) {
    override fun equals(other: Any?): Boolean = other is Record && id == other.id && userId == other.userId && timestamp == other.timestamp && vector.contentEquals(other.vector)
    override fun hashCode(): Int = 31 * (31 * (31 * id.hashCode() + userId.hashCode()) + timestamp.hashCode()) + vector.contentHashCode()
}

/** Optional user equality and lower-inclusive, upper-exclusive timestamp bounds. */
public data class Filter(
    public val userId: ULong? = null,
    public val timestampLower: Long? = null,
    public val timestampUpper: Long? = null,
)

public data class SearchResult(public val id: ULong, public val distance: Float)

public data class ExecutionReport(
    public val operationId: ULong,
    public val requestedBackend: String,
    public val actualBackend: String,
    public val algorithm: String,
    public val filterExecution: String,
    public val indexGeneration: ULong,
    public val rebuilt: Boolean,
    public val routingReason: String?,
    public val fallbackReason: String?,
    public val totalDurationNs: ULong,
    public val lockWaitNs: ULong,
    public val eligibleRows: ULong?,
    public val uploadBytes: ULong?,
    public val readbackBytes: ULong?,
    public val allocationBytes: ULong?,
    public val dispatchCount: UInt?,
    public val candidates: ULong?,
    public val batchSize: Int,
)

public data class SearchResponse(public val results: List<SearchResult>, public val report: ExecutionReport)

public data class CollectionStats(
    public val dimension: Int,
    public val rows: Int,
    public val liveRows: Int,
    public val generation: ULong,
    public val preparedGeneration: ULong?,
    public val durableGeneration: ULong?,
    public val recoveredInterruptedWrite: Boolean,
    public val closed: Boolean,
)

private interface QenloNative : Library {
    fun qenlo_collection_new(dimension: Long): Pointer?
    fun qenlo_collection_new_configured(dimension: Long, backend: Int, gpuFilterMode: Int, gpuAllocationBudgetBytes: Long): Pointer?
    fun qenlo_collection_create(path: String, dimension: Long): Pointer?
    fun qenlo_collection_create_configured(path: String, dimension: Long, backend: Int, gpuFilterMode: Int, gpuAllocationBudgetBytes: Long): Pointer?
    fun qenlo_collection_open(path: String, dimension: Long): Pointer?
    fun qenlo_collection_open_configured(path: String, dimension: Long, backend: Int, gpuFilterMode: Int, gpuAllocationBudgetBytes: Long): Pointer?
    fun qenlo_collection_import_qn(path: String, dimension: Long): Pointer?
    fun qenlo_collection_import_qn_configured(path: String, dimension: Long, backend: Int, gpuFilterMode: Int, gpuAllocationBudgetBytes: Long): Pointer?
    fun qenlo_add(handle: Pointer, id: Long, userId: Long, timestamp: Long, vector: FloatArray, vectorLen: Long): Int
    fun qenlo_add_batch(handle: Pointer, ids: LongArray, userIds: LongArray, timestamps: LongArray, vectors: FloatArray, rows: Long, dimension: Long): Int
    fun qenlo_delete(handle: Pointer, id: Long): Int
    fun qenlo_delete_batch(handle: Pointer, ids: LongArray, rows: Long): Int
    fun qenlo_search(handle: Pointer, query: FloatArray, queryLen: Long, hasUserId: Byte, userId: Long, hasLower: Byte, lower: Long, hasUpper: Byte, upper: Long, k: Long): Pointer?
    fun qenlo_stats(handle: Pointer): Pointer?
    fun qenlo_export_qn(handle: Pointer, path: String): Int
    fun qenlo_flush(handle: Pointer): Int
    fun qenlo_close(handle: Pointer): Int
    fun qenlo_collection_free(handle: Pointer)
    fun qenlo_last_error(): Pointer?
    fun qenlo_string_free(value: Pointer)
}

private object NativeApi {
    val value: QenloNative by lazy {
        val explicit = System.getenv("QENLO_LIBRARY_PATH")?.let { Path.of(it) }
        val library = when {
            explicit?.toFile()?.isFile == true -> explicit.toAbsolutePath().toString()
            explicit != null -> {
                NativeLibrary.addSearchPath("qenlo_ffi", explicit.toAbsolutePath().toString())
                "qenlo_ffi"
            }
            else -> bundledLibrary()?.toString() ?: "qenlo_ffi"
        }
        Native.load(library, QenloNative::class.java)
    }

    private fun bundledLibrary(): Path? {
        val os = System.getProperty("os.name").lowercase()
        val arch = System.getProperty("os.arch").lowercase()
        val platform = when {
            os.contains("win") && arch in setOf("amd64", "x86_64") -> "windows-x64"
            os.contains("linux") && arch in setOf("amd64", "x86_64") -> "linux-x64"
            os.contains("mac") && arch in setOf("aarch64", "arm64") -> "darwin-arm64"
            else -> return null
        }
        val file = when {
            os.contains("win") -> "qenlo_ffi.dll"
            os.contains("mac") -> "libqenlo_ffi.dylib"
            else -> "libqenlo_ffi.so"
        }
        val resource = "/qenlo/$platform/$file"
        val input = NativeApi::class.java.getResourceAsStream(resource) ?: return null
        val directory = Files.createTempDirectory("qenlo-native-").also { it.toFile().deleteOnExit() }
        val target = directory.resolve(file)
        input.use { Files.copy(it, target, StandardCopyOption.REPLACE_EXISTING) }
        target.toFile().deleteOnExit()
        return target
    }
}

@Serializable
private data class WireHit(val id: String, val distance: Float)

@Serializable
private data class WireReport(
    @SerialName("operation_id") val operationId: String,
    @SerialName("requested_backend") val requestedBackend: String,
    @SerialName("actual_backend") val actualBackend: String,
    val algorithm: String,
    @SerialName("filter_execution") val filterExecution: String,
    @SerialName("index_generation") val indexGeneration: String,
    val rebuilt: Boolean,
    @SerialName("routing_reason") val routingReason: String?,
    @SerialName("fallback_reason") val fallbackReason: String?,
    @SerialName("total_duration_ns") val totalDurationNs: String,
    @SerialName("lock_wait_ns") val lockWaitNs: String,
    @SerialName("eligible_rows") val eligibleRows: String?,
    @SerialName("upload_bytes") val uploadBytes: String?,
    @SerialName("readback_bytes") val readbackBytes: String?,
    @SerialName("allocation_bytes") val allocationBytes: String?,
    @SerialName("dispatch_count") val dispatchCount: UInt?,
    val candidates: String?,
    @SerialName("batch_size") val batchSize: Int,
)

@Serializable
private data class WireSearch(val results: List<WireHit>, val report: WireReport)

@Serializable
private data class WireStats(
    val dimension: Int,
    val rows: Int,
    @SerialName("live_rows") val liveRows: Int,
    val generation: String,
    @SerialName("prepared_generation") val preparedGeneration: String?,
    @SerialName("durable_generation") val durableGeneration: String?,
    @SerialName("recovered_interrupted_write") val recoveredInterruptedWrite: Boolean,
    val closed: Boolean,
)

/** An owned Qenlo collection. Close explicitly or use [use]. */
public class QenloCollection private constructor(handle: Pointer?, public val dimension: Int) : AutoCloseable {
    private val lock: ReentrantReadWriteLock = ReentrantReadWriteLock()
    private var handle: Pointer? = handle ?: throw QenloException(lastError())

    public companion object {
        public fun memory(dimension: Int, options: CollectionOptions = CollectionOptions()): QenloCollection {
            require(dimension > 0) { "dimension must be positive" }
            return QenloCollection(
                NativeApi.value.qenlo_collection_new_configured(
                    dimension.toLong(), options.backend.nativeValue, options.gpuFilterMode.nativeValue,
                    options.gpuAllocationBudgetBytes.toLong(),
                ),
                dimension,
            )
        }

        public fun create(path: Path, dimension: Int, options: CollectionOptions = CollectionOptions()): QenloCollection {
            require(dimension > 0) { "dimension must be positive" }
            return QenloCollection(
                NativeApi.value.qenlo_collection_create_configured(
                    path.toString(), dimension.toLong(), options.backend.nativeValue,
                    options.gpuFilterMode.nativeValue, options.gpuAllocationBudgetBytes.toLong(),
                ),
                dimension,
            )
        }

        public fun open(path: Path, dimension: Int, options: CollectionOptions = CollectionOptions()): QenloCollection {
            require(dimension > 0) { "dimension must be positive" }
            return QenloCollection(
                NativeApi.value.qenlo_collection_open_configured(
                    path.toString(), dimension.toLong(), options.backend.nativeValue,
                    options.gpuFilterMode.nativeValue, options.gpuAllocationBudgetBytes.toLong(),
                ),
                dimension,
            )
        }

        public fun importQn(path: Path, dimension: Int, options: CollectionOptions = CollectionOptions()): QenloCollection {
            require(dimension > 0) { "dimension must be positive" }
            return QenloCollection(
                NativeApi.value.qenlo_collection_import_qn_configured(
                    path.toString(), dimension.toLong(), options.backend.nativeValue,
                    options.gpuFilterMode.nativeValue, options.gpuAllocationBudgetBytes.toLong(),
                ),
                dimension,
            )
        }

        private fun lastError(): String = takeString(NativeApi.value.qenlo_last_error(), false)

        private fun takeString(pointer: Pointer?, failOnNull: Boolean = true): String {
            if (pointer == null) {
                if (failOnNull) throw QenloException(lastError())
                return "unknown Qenlo native error"
            }
            return try { pointer.getString(0, Charsets.UTF_8.name()) } finally { NativeApi.value.qenlo_string_free(pointer) }
        }
    }

    private fun openHandle(): Pointer = handle ?: throw QenloException("collection is closed")
    private fun check(status: Int) { if (status != 0) throw QenloException(lastError()) }
    private fun vector(value: FloatArray): FloatArray {
        require(value.size == dimension) { "expected vector dimension $dimension, got ${value.size}" }
        return value
    }

    public fun add(record: Record): Unit = lock.read {
        check(NativeApi.value.qenlo_add(openHandle(), record.id.toLong(), record.userId.toLong(), record.timestamp, vector(record.vector), dimension.toLong()))
    }

    public fun addBatch(records: List<Record>): Unit = lock.read {
        if (records.isEmpty()) return@read
        val ids = LongArray(records.size)
        val users = LongArray(records.size)
        val timestamps = LongArray(records.size)
        val vectors = FloatArray(records.size * dimension)
        records.forEachIndexed { row, record ->
            ids[row] = record.id.toLong(); users[row] = record.userId.toLong(); timestamps[row] = record.timestamp
            vector(record.vector).copyInto(vectors, row * dimension)
        }
        check(NativeApi.value.qenlo_add_batch(openHandle(), ids, users, timestamps, vectors, records.size.toLong(), dimension.toLong()))
    }

    public fun delete(id: ULong): Unit = lock.read { check(NativeApi.value.qenlo_delete(openHandle(), id.toLong())) }
    public fun deleteBatch(ids: List<ULong>): Unit = lock.read {
        if (ids.isNotEmpty()) check(NativeApi.value.qenlo_delete_batch(openHandle(), ids.map(ULong::toLong).toLongArray(), ids.size.toLong()))
    }

    public fun search(query: FloatArray, filter: Filter = Filter(), k: Int = 10): SearchResponse = lock.read {
        require(k in 1..64) { "k must be in 1..=64" }
        val pointer = NativeApi.value.qenlo_search(openHandle(), vector(query), dimension.toLong(), (if (filter.userId != null) 1 else 0).toByte(), filter.userId?.toLong() ?: 0, (if (filter.timestampLower != null) 1 else 0).toByte(), filter.timestampLower ?: 0, (if (filter.timestampUpper != null) 1 else 0).toByte(), filter.timestampUpper ?: 0, k.toLong())
        val wire = Json.decodeFromString<WireSearch>(takeString(pointer))
        SearchResponse(
            wire.results.map { SearchResult(it.id.toULong(), it.distance) },
            wire.report.run { ExecutionReport(operationId.toULong(), requestedBackend, actualBackend, algorithm, filterExecution, indexGeneration.toULong(), rebuilt, routingReason, fallbackReason, totalDurationNs.toULong(), lockWaitNs.toULong(), eligibleRows?.toULong(), uploadBytes?.toULong(), readbackBytes?.toULong(), allocationBytes?.toULong(), dispatchCount, candidates?.toULong(), batchSize) },
        )
    }

    public fun stats(): CollectionStats = lock.read {
        val wire = Json.decodeFromString<WireStats>(takeString(NativeApi.value.qenlo_stats(openHandle())))
        CollectionStats(wire.dimension, wire.rows, wire.liveRows, wire.generation.toULong(), wire.preparedGeneration?.toULong(), wire.durableGeneration?.toULong(), wire.recoveredInterruptedWrite, wire.closed)
    }

    public fun flush(): Unit = lock.read { check(NativeApi.value.qenlo_flush(openHandle())) }

    public fun exportQn(path: Path): Unit = lock.read {
        check(NativeApi.value.qenlo_export_qn(openHandle(), path.toString()))
    }

    override fun close(): Unit = lock.write {
        handle?.let { value ->
            handle = null
            val status = NativeApi.value.qenlo_close(value)
            NativeApi.value.qenlo_collection_free(value)
            check(status)
        }
    }
}
